### ISSUE TITLE
Reducing the "keep DL" magic down to two bytes

### ISSUE THREAD

**Issue by cyrozap** [orangecms/sg_boot#2]
Hi, by sheer coincidence within the past week I'd started working on a utility in Rust to do basically the same thing `sg_boot` does, and just today discovered your `sg_boot` program. I've also been doing some reverse engineering of the SG200x boot ROM and while doing so found a fun little opportunity for simplifying the USB boot process that you might be interested in.

In case you're not aware, the 128 "magic bytes" needed to keep the chip in USB DL mode can be reduced down to just two bytes, which eliminates having to carry around a binary blob (with unknown licensing implications) in every utility that communicates with these chips. This optimization is possible because the magic bytes, which can be of any length, only exist to have [a CRC-16 calculated over them](https://github.com/sophgo/fsbl/blob/629f0d766bd527c4624ba277ac2bc27c16d840ba/plat/cv180x/usb/usb_tty.c#L307), which is then checked to match the value [0xC283](https://github.com/sophgo/fsbl/blob/629f0d766bd527c4624ba277ac2bc27c16d840ba/plat/cv180x/usb/usb_tty.c#L309), so any sequence of bytes that has a CRC-16 value equal to 0xC283 will work.

After a quick brute-force using some unoptimized Python code (the command returned nearly instantly), I discovered that the byte sequence `0x2c, 0x08` generates the correct CRC value, and I've confirmed that it works with real hardware (both the CV1800B and SG2002, and I assume the SG2000 as well). To confirm for yourself that the CRC values are the same, you can compare the [CRC of the original blob](https://crccalc.com/?crc=0800001400c018d50006a0d2401018d5400000944000003500000014a03f00100800001400c018d50006a0d2401018d5400000944000003500000014a03f001000113ed5000c40b200111ed55f111ed500103ed5a10080d20000218a00101ed51000001400c01cd5e07f86d240111cd500103cd5a10080d20000218a00101cd5&method=CRC-16/XMODEM&datatype=hex&outtype=0) with [that of the two-byte solution](https://crccalc.com/?crc=0x2c,%200x08&method=CRC-16/XMODEM&datatype=hex&outtype=0).

tl;dr: You can replace [this line](https://github.com/orangecms/sg_boot/blob/749617bfc96dae64b43c3d37534d78e27b77e9a9/src/protocol.rs#L3) with something like `const NO_MAGIC: [u8; 2] = [0x2c, 0x08];` and you won't have to use `nomagic.bin` anymore.

**Comment by cyrozap** [comment]
Also, FWIW, the `DUMMY_ADDR` is only necessary [when the length is zero](https://github.com/sophgo/fsbl/blob/629f0d766bd527c4624ba277ac2bc27c16d840ba/plat/cv180x/usb/usb_tty.c#L278), so if you always set the total length to at least the length of the header (which [you appear to be doing](https://github.com/orangecms/sg_boot/blob/749617bfc96dae64b43c3d37534d78e27b77e9a9/src/protocol.rs#L51)) then it's unnecessary to ever set the address to `DUMMY_ADDR` (it can be left as zero when it's unused).

**Comment by pftbest** [comment]
Tried your pull request, works great with MilkV Duo S (SG2000). It's nice to have one less binary in the chain, thank you!

**Comment by orangecms** [comment]
Oh nice, thank you! 


And sorry, I haven't seen this earlier due to missing notifications, I suppose. =)
